command/daemon.go

```diff
@@ -729,3 +752,83 @@ func createDatastore(cfg config.Datastore) (datastore.Batching, error) {
 	}
 	return leveldb.NewDatastore(dataStorePath, nil)
 }
+
+func updateDatastore(ctx context.Context, ds datastore.Batching) error {
+	dsVerKey := datastore.NewKey(dsInfoPrefix + dsVersionKey)
+	curVerData, err := ds.Get(ctx, dsVerKey)
+	if err != nil && !errors.Is(err, datastore.ErrNotFound) {
+		return fmt.Errorf("cannot check datastore: %w", err)
+	}
+	var curVer string
+	if len(curVerData) != 0 {
+		curVer = string(curVerData)
+	}
+	if curVer == dsVersion {
+		return nil
+	}
+	if curVer > dsVersion {
+		return fmt.Errorf("unknown datastore verssion: %s", curVer)
+	}
+
+	log.Infof("Updating datastore to version %s", dsVersion)
+
+	q := query.Query{
+		KeysOnly: true,
+	}
+	results, err := ds.Query(ctx, q)
+	if err != nil {
+		return fmt.Errorf("cannot query datastore: %w", err)
+	}
+	defer results.Close()
+
+	batch, err := ds.Batch(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot create datastore batch: %w", err)
+	}
+
+	var cidCount, dtKeyCount, writeCount int
+	for result := range results.Next() {
+		if writeCount >= 500000 {
+			writeCount = 0
+			if err = batch.Commit(ctx); err != nil {
+				return fmt.Errorf("cannot commit to datastore: %w", err)
+			}
+			log.Infow("Datastore update removed data", "dtSessions", dtKeyCount, "cids", cidCount)
+		}
+		if result.Error != nil {
+			return fmt.Errorf("cannot read query result from datastore: %w", result.Error)
+		}
+		ent := result.Entry
+		key := ent.Key[1:]
```

Comment:
is it possible for `ent.Key` to be empty / for this to panic?

Reply:
It is impossible, at least with leveldb:
https://github.com/ipfs/go-ds-leveldb/blob/master/datastore.go#L135-L136

Will handle this case anyway just in case there is some implementation that does have same behavior.

Reply:
Done